### PR TITLE
 Return a list of VM sizes in calculate_vm_cloud_properties

### PR DESCRIPTION
**Pull request by bingosummer**
### Changelog

* Refine RECOMMENDED_VM_SIZES in calculate_vm_cloud_properties
    * Include instance types which support Premium storage
* Use the instance types which support Premium storage if the attached disk is premium disk

### CODE REVIEW

File: src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb [hunk 1]
```diff
@@ -19,23 +19,57 @@ def initialize(azure_config, registry_endpoint, disk_manager, disk_manager2, azu
...
+        availability_set_name = _get_availability_set_name(vm_props, env)
+        if vm_props.instance_type.nil?
+          calculated_instance_types = vm_props.instance_types.dup
+          @logger.debug("The cloud property 'instance_type' is nil. Will select one from '#{calculated_instance_types}', which are returned by calculate_vm_cloud_properties.")
+
+          is_persistent_disk_premium = false
+          disk_cids&.each do |disk_cid|
+            disk_id = DiskId.parse(disk_cid, @azure_config.resource_group_name)
+            disk = @disk_manager2.get_data_disk(disk_id)
+            if disk[:sku_tier] == SKU_TIER_PREMIUM
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb]: @disk_manager2.get_data_disk(disk_id) can return nil if the disk is not a managed disk, so disk[:sku_tier] will raise an error.

File: src/bosh_azure_cpi/lib/cloud/azure/models/vm_cloud_props.rb [hunk 2]
```diff
@@ -1,60 +1,31 @@
...
-      @ip_forwarding = vm_properties['ip_forwarding']
-      @accelerated_networking = vm_properties['accelerated_networking']
-      @assign_dynamic_public_ip = vm_properties['assign_dynamic_public_ip']
-      @load_balancer = vm_properties['load_balancer']
-      @application_gateway = vm_properties['application_gateway']
-
-      @application_security_groups = vm_properties['application_security_groups']
-      @security_group = vm_properties['security_group']
+      @instance_types = vm_properties['instance_types']
+      cloud_error("The cloud property 'instance_type' is missing. Alternatively you can specify 'vm_resources' key. https://bosh.io/docs/manifest-v2/#instance-groups") if @instance_type.nil? && (@instance_types.nil? || @instance_types.empty?)
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/models/vm_cloud_props.rb]: this log is a little confusing, 'instance_types' is not mentioned.

File: src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb (hunk 1 above)
Comment (bingosummer) [reply src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb]: Good catch.

File: src/bosh_azure_cpi/lib/cloud/azure/models/vm_cloud_props.rb (hunk 2 above)
Comment (bingosummer) [reply src/bosh_azure_cpi/lib/cloud/azure/models/vm_cloud_props.rb]: Users don't know about `instance_groups`. They only known `vm_resources`. But yes, the log is not clear. Will try to refine it.

File: src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb [hunk 3]
```diff
@@ -19,23 +19,57 @@ def initialize(azure_config, registry_endpoint, disk_manager, disk_manager2, azu
...
+            @logger.debug("The disk that created VM will be attached is a Premium disk. One of the instance types '#{calculated_instance_types}' which support Premium storage will be used.")
+          end
 
-      _check_resource_group(resource_group_name, location)
+          availability_set = @azure_client.get_availability_set_by_name(resource_group_name, availability_set_name)
+          unless availability_set.nil?
+            available_vm_sizes = @azure_client.list_available_virtual_machine_sizes_by_availability_set(resource_group_name, availability_set_name)
+            available_vm_sizes = available_vm_sizes.map { |available_vm_size| available_vm_size[:name] }
+            @logger.debug("The availability set '#{availability_set_name}' supports VM sizes '#{available_vm_sizes}'")
+            calculated_instance_types = calculated_instance_types.select do |calculated_instance_type|
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb]: calculated_instance_types = calculated_instance_types & available_vm_sizes

see https://stackoverflow.com/questions/5678108/how-can-i-get-the-intersection-union-and-subset-of-arrays-in-ruby

File: src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb [hunk 4]
```diff
@@ -421,6 +422,28 @@ def list_available_virtual_machine_sizes(location)
...
+    #      https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/Microsoft.Compute/stable/2018-04-01/compute.json
+    #
+    def list_available_virtual_machine_sizes_by_availability_set(resource_group_name, availability_set_name)
+      vm_sizes = []
+      url = rest_api_url(REST_API_PROVIDER_COMPUTE, REST_API_AVAILABILITY_SETS, resource_group_name: resource_group_name, name: availability_set_name, others: REST_API_VM_SIZES)
+      result = get_resource_by_id(url)
+
+      unless result.nil? || result['value'].nil?
+        result['value'].each do |value|
+          vm_size = parse_vm_size(value)
```
Comment (gossion) [inline src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb]: it is ok to put these two lines into one.

File: src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb (hunk 3 above)
Comment (bingosummer) [reply src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb]: Great suggestion.

**Comment by bingosummer** [comment]
Closing it for now. Will re-submit a new PR after rebasing commits and refining the codes.

**Pull request closed without merging**